nvdaHelper/local/screenCurtain.cpp

```diff
@@ -0,0 +1,186 @@
+		screenArea = screenWidth * screenHeight,
+		bytesWritten;
+	UINT histogramSize;
+	DWORD diScreenshotSize;
+	HWND desktopWnd;
+	BITMAP ddScreenshot;
+	HGDIOBJ oldObj;
+	BITMAPINFOHEADER diScreenshotHeader;
+	bool bStatus ;
+	Status sStatus;
+
+	// The desktop window covers the entire virtual screen.
+	desktopWnd = GetDesktopWindow();
+	if (desktopWnd == NULL) {
+		LOG_ERROR(L"Failed to get handle for desktop window.");
+		return false;
+	}
+	wil::unique_hdc_window desktopDc(GetDC(desktopWnd));
+	if (!desktopDc.is_valid()) {
+		LOG_ERROR(L"Failed to get device context for desktop.");
+		return false;
+	}
+	wil::unique_hdc captureDc(CreateCompatibleDC(desktopDc.get()));
+	if (!captureDc.is_valid()) {
+		LOG_ERROR("Failed to create compatible device context.");
+		return false;
+	}
+	wil::unique_hbitmap captureBitmap(CreateCompatibleBitmap(desktopDc.get(), screenWidth, screenHeight));
+	if (captureBitmap == NULL) {
+		LOG_ERROR(L"Failed to create compatible bitmap.");
+		return false;
+	}
+
+	// Set captureDc to draw to captureBitmap.
+	oldObj = SelectObject(captureDc.get(), captureBitmap.get());
+	if (oldObj == NULL) {
+		LOG_ERROR("Failed to select capture bitmap into capture device context.");
+		return false;
+	}
+
+	// Replace the contents of captureDc with those of desktopDc
+	bStatus = BitBlt(
+		// Destination device context
+		captureDc.get(),
+		// Top left of destination
+		0, 0,
+		// Size of source and destination
+		screenWidth, screenHeight,
+		// Source device context
+		desktopDc.get(),
+		// Top left of source
+		screenOriginX, screenOriginY,
+		// Raster operation to perform.
+		// In this case, replace destination with source.
+		SRCCOPY);
+	// Restore captureDC for safety.
+	SelectObject(captureDc.get(), oldObj);
+	if (!bStatus) {
+		LOG_ERROR("Failed to bit blit desktop device context to capture device context. Error #" << GetLastError());
+		return false;
+	}
+
+	// Get properties of captureBitmap
+	bytesWritten = GetObject(captureBitmap.get(), sizeof(BITMAP), &ddScreenshot);
+	if (bytesWritten == 0) {
+		LOG_ERROR("Failed to get bitmap metadata.");
+		return false;
+	}
+
+	// And use them to set the properties of the device independent bitmap.
+	diScreenshotHeader.biSize = sizeof(BITMAPINFOHEADER);
+	diScreenshotHeader.biWidth = ddScreenshot.bmWidth;
+	diScreenshotHeader.biHeight = ddScreenshot.bmHeight;
+	diScreenshotHeader.biPlanes = 1;			  // Can only ever be 1
+	diScreenshotHeader.biBitCount = 32;		  // High byte unused
+	diScreenshotHeader.biCompression = BI_RGB; // Uncompressed
+	diScreenshotHeader.biSizeImage = 0;		  // Unneeded as uncompressed
+	diScreenshotHeader.biXPelsPerMeter = 0;
+	diScreenshotHeader.biYPelsPerMeter = 0;
+	diScreenshotHeader.biClrUsed = 0;
+	diScreenshotHeader.biClrImportant = 0; // All colours are needed
+
+	// Calculate the size (in bytes) of the DIB.
+	diScreenshotSize = ((ddScreenshot.bmWidth * diScreenshotHeader.biBitCount + 31) / 32) * 4 * ddScreenshot.bmHeight;
```

Comment:
why 31?

Reply:
Because each scan line has to be a multiple of 32 bits in length. If the scan line is less than 32 bits long, the end bits are padding, and their value does not matter.
Dividing ints in C gives us the quotient, so we need to make sure if we have a scanline with padding bits, we'll capture those bits. If the scan line fits perfectly into 32 bits, however, adding 31 will not tip us over into a larger value.
```suggestion
	// Each scan line must be a multiple of 32bits long, including padding if necessary.
	// So add 31 to push us to that boundary if padding is needed.
	// If the scan line doesn't need any padding, adding 31 will make no difference,
	// since (32n + 31) / 32 = n R 31.
	diScreenshotSize = ((ddScreenshot.bmWidth * diScreenshotHeader.biBitCount + 31) / 32) * 4 * ddScreenshot.bmHeight;
```